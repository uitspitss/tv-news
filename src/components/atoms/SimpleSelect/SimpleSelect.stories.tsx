import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { SimpleSelect, SimpleSelectProps } from './SimpleSelect';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: SimpleSelect,
} as Meta;

const Template: Story<SimpleSelectProps> = (args) => <SimpleSelect {...args} />;

export const Default = Template.bind({});
Default.args = {
  name: 'select',
  options: [
    { value: 'op1', label: 'label1' },
    { value: 'op2', label: 'label2' },
    { value: 'op3', label: 'label3' },
  ],
};
