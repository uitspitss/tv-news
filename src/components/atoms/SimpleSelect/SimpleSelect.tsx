import React, { memo, VFC, ComponentPropsWithRef } from 'react';
import tw, { styled } from 'twin.macro';

type OptionType = {
  value: string | number;
  label: string;
};

export type SimpleSelectProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * options
   */
  options: Readonly<OptionType[]>;
  /**
   * label
   */
  label: string;
  /**
   * error message
   */
  error?: string;
} & ComponentPropsWithRef<'select'>;

type Props = SimpleSelectProps;

const Component: VFC<Props> = (props) => {
  const { className, name, label, options = [], error, ...rest } = props;

  return (
    <div className={className}>
      <label htmlFor={name}>{label}</label>
      <select
        name={name}
        id={name}
        aria-label={label}
        aria-invalid={error ? 'true' : 'false'}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...rest}
      >
        {options.map((option) => (
          <option key={option.label} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      {error && (
        <span className="error" role="alert">
          {error}
        </span>
      )}
    </div>
  );
};

const StyledComponent = styled(Component)`
  & > {
    select {
      ${tw`w-full bg-gray-100 rounded border border-gray-400 focus:outline-none focus:border-indigo-500`}
    }

    .error {
      ${tw`text-red-500`}
    }
  }
`;

/**
 * SimpleSelect
 */
export const SimpleSelect = memo((props: SimpleSelectProps) => {
  return <StyledComponent {...props} />;
});
