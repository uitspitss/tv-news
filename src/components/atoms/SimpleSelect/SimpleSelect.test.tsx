import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { SimpleSelect } from './SimpleSelect';

describe('atoms/SimpleSelect', () => {
  const options = [
    { value: 'op1 value', label: 'op1 label' },
    { value: 'op2 value', label: 'op2 label' },
    { value: 'op3 value', label: 'op3 label' },
  ];

  test('snapshot test', () => {
    const { container } = render(
      <SimpleSelect
        name="select-name"
        label="select label"
        placeholder="select placeholder"
        options={options}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      render(
        <SimpleSelect
          name="select-name"
          label="select label"
          placeholder="select placeholder"
          options={options}
        />,
      );
    });

    test('initial view', () => {
      expect(screen.getByLabelText('select label')).toBeInTheDocument();
      expect(screen.getByText('select label')).toBeInTheDocument();
    });

    test('select, then selected', () => {
      fireEvent.select(screen.getByLabelText('select label'), 'op1 label');
      expect(screen.getByLabelText('select label')).toHaveValue('op1 value');
    });
  });
});
