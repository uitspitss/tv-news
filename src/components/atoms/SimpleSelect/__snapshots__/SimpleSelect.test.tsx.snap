// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/SimpleSelect snapshot test 1`] = `
.emotion-0>select {
  width: 100%;
  --tw-bg-opacity: 1;
  background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
}

.emotion-0>select:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(99, 102, 241, var(--tw-border-opacity));
}

.emotion-0>.error {
  --tw-text-opacity: 1;
  color: rgba(239, 68, 68, var(--tw-text-opacity));
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <label
      for="select-name"
    >
      select label
    </label>
    <select
      aria-invalid="false"
      aria-label="select label"
      id="select-name"
      name="select-name"
      placeholder="select placeholder"
    >
      <option
        value="op1 value"
      >
        op1 label
      </option>
      <option
        value="op2 value"
      >
        op2 label
      </option>
      <option
        value="op3 value"
      >
        op3 label
      </option>
    </select>
  </div>
</div>
`;
